clients/admin-ui/src/features/common/CheckboxTree.tsx

```diff
@@ -1,5 +1,56 @@
-import { Box, Checkbox } from "@fidesui/react";
-import { Fragment, ReactNode } from "react";
+/**
+ * This CheckboxTree does a few opinionated things to facilitate DataCategory selection
+ *   * Indeterminate parents when not all children are selected
+ *   * Separate "selected" and "checked" states, since only the most specific child is "selected".
+ *     * However, the parents are not "selected"—they will either be "indeterminate", or "checked" depending
+ *       on if the siblings are also "selected"
+ *   * "Selected" children render expanded up until the child
+ */
+
+import { Box, Checkbox, IconButton } from "@fidesui/react";
+import { Fragment, ReactNode, useEffect, useState } from "react";
+
+import { ArrowDownLineIcon, ArrowUpLineIcon } from "./Icon";
+
+export const getAncestorsAndCurrent = (nodeName: string) => {
+  const splitNames = nodeName.split(".");
+  const ancestors: string[] = [];
+  splitNames.forEach((name) => {
+    const parent =
+      ancestors.length > 0 ? ancestors[ancestors.length - 1] : null;
+    if (!parent) {
+      ancestors.push(name);
+    } else {
+      ancestors.push(`${parent}.${name}`);
+    }
+  });
+  return ancestors;
+};
+
+export const ancestorIsSelected = (selected: string[], nodeName: string) => {
+  const ancestors = getAncestorsAndCurrent(nodeName).filter(
+    (a) => a !== nodeName
+  );
+  const intersection = selected.filter((s) => ancestors.includes(s));
+  return intersection.length > 0;
+};
+
+export const getDescendantsAndCurrent = (
+  nodes: CheckboxNode[],
+  nodeName: string,
+  result?: CheckboxNode[]
+) => {
+  const descendants: CheckboxNode[] = result ?? [];
+  nodes.forEach((node) => {
+    if (node.children) {
+      getDescendantsAndCurrent(node.children, nodeName, descendants);
+    }
+    if (node.value.startsWith(nodeName)) {
+      descendants.push(node);
```

Comment:
Same thing here with descendants - I think it's stylistically atypical to use push/pop in a React ecosystem, not that it doesn't work.